File: Scoping/02_Event_Extract.R
######
# Adam Griffin, 2020-02-13
#
# Event extraction from daily flow. Practiced on one 30-year period of data.
# For aquaCAT, Project 07441.
# 
# Created ABG 2020-02-13
#
#####

##### SETUP ##### --------------------------------------------------------

#### Packages -------------------------------------------------------
library(ncdf4)
library(fields)
library(raster)


#### Data -----------------------------------------------------------


## Read in netCDF ------

ncname_lin <-"/prj/ResilRiskInds/hmfg2g_runs/baseline_qx_setting/dmflow_out.nc"
ncname_win <- "K:/hmfg2g_runs/baseline_qx_setting/dmflow_out.nc"  
# This file is ~36GB on the linux server.

if(substr(osVersion,1,3)=="Win"){
  ncname <- ncname_win
}else{
  ncname <- ncname_lin
}
ncin <- nc_open(ncname, readunlim=FALSE)
print("ncin open")

# Subset for testing
istart <- 1
Di <- 50 
jstart <- 1
Dj <- 50
tstart <- 1

## Restrict to river network ---------

# `NA` is sea pixels, `-1` is not on the river network.
Dt <- 1
ncwide <- ncvar_get(ncin, "dmflow",
                    start=c(1,1,tstart), count=c(-1, -1, Dt))
print("ncwide get")

HasData <- which(apply(ncwide, c(1,2),
                       function(v){sum(v[!is.na(v)] > -1) == Dt}),
                 arr.ind=T)
NH <- nrow(HasData)
# HasData is important for gluing the events back in at the other end 
# of the pipeline.


### EXTRA IMPORTANT SHAPES -------------------------------------------------
HD <- raster(x=ncname_win, layer=1)  # gets extent etc. automatically
HD[HD < 0] <- NA  # sets "land" to "ocean"
HD[HD >= 0] <- 1 # sets "rivers" to 1

writeRaster(HD, "./Scoping/InterimData/RiverNetwork.asc", format='ascii')

HX <- raster(x=ncname_win, layer=1)  # gets extent etc. automatically
HX[HX >= -1] <- -1  # sets rivers to "land"
plot(HX)
writeRaster(HX, "./Scoping/InterimData/GBshape.asc", format="ascii")


## Event extraction ------------
qpot5_grid <- raster(x=ncname_win, layer=1)

qpot5_grid[!is.na(qpot5_grid) & qpot5_grid > -1] <- 0

qpot_extract <- function(n){
  i <- HasData[n,1]
  j <- HasData[n,2]
  tslice <- ncvar_get(ncin, varid="dmflow",
                      start=c(i, j,  1),
                      count=c(1, 1, 10957))
  qpot5 <- quantile(as.vector(tslice), prob=c(1 - (5/365)))
  qpot5_grid[i,j] <- qpot5
}

qpot_extract_full <- function(ncfile, varid, quant, step=50){
  
  qpot_grid <- raster(x=ncfile, layer=1)
  di <- seq(1, dim(qpot_grid)[1], by=step)
  dj <- seq(1, dim(qpot_grid)[2], by=step)
  pb <- txtProgressBar(0, length(di)*length(dj))
  k <- 0
  for(i in di){
    for(j in dj){
      tslice <- ncvar_get(ncin, varid="dmflow",
                          start=c(i, j,  1),
                          count=c(step, step, -1))
      qpot <- apply(X=tslice, c(1,2), FUN=quantile, probs=quant, na.rm=T,
                    names=FALSE)
      qpot_grid[i+(0:(step-1)), j+(0:(step-1))] <- qpot
      k <- k+1
      setTxtProgressBar(pb, k)
    }
  }
  close(pb)
  return(qpot_grid)
  
}

QP99 <- qpot_extract_full(ncname_win, "dmflow", 0.99, step=50)



for(n in 1:NH){
  qpot_extract(n)
  if(n %% 1 == 0){
    print("Expected time to end = ")
    print(((NH-n)/n)*difftime(Sys.time(),ST.old, unit="hours"))
  }
}

save(qpot5_grid, file="./InterimData/qpot5_grid.RDa")

File: Scoping/02_Event_Extract.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
 Adam Griffin, 2020-02-21

 Event extraction from daily flow. Practiced on one 30-year period of data.
 Rewritten for linux to reduce read-write times for netCDF.

 For aquaCAT, Project 07441.
 
Created ABG 2020-02-21
Updated ABG 2020-03-05 - continued development
"""

from netCDF4 import Dataset
import numpy as np
import pandas as pd
import numpy.ma as ma
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm

nc_f = "/prj/ResilRiskInds/hmfg2g_runs/baseline_qx_setting/dmflow_out.nc"
mainfolder = "/users/sgsys/adagri/aquacat"

ncin = Dataset(nc_f, 'r')
ncshape = ncin.variables['dmflow'].shape  #(10957, 1250, 700)
flow_wide = ncin.variables['dmflow'][0,:,:]

qPOT5 = flow_wide
dx = 50
events_per_year = 5
flow_thres = 1 - (events_per_year/365)
inun_thres = 0.01

for i in range(0, ncshape[1]+1, dx):
    for j in range(0, ncshape[2]+1, dx):
        if flow_wide.mask[i:min(i+dx, ncshape[1]), j:min(j+dx, ncshape[2])].all():
            # mask is true if masking -9999, skip the chunk if all masked
            continue
        else:
            print(str(i) + " " + str(j))
            # extract dmflow in chunks to reduce read-write time
            flow_sub = ncin.variables['dmflow'][0:ncshape[0],
                                                i:min(i+dx, ncshape[1]),
                                                j:min(j+dx, ncshape[2])]
            # compute quantile
            q5 = np.quantile(flow_sub, q=flow_thres, axis=0)
            qPOT5[i:min(i+dx, ncshape[1]), j:min(j+dx, ncshape[2])] = q5
            # preserve "non-river" value of -1
            qPOT5[np.where(flow_wide[i:min(i+dx, ncshape[1]),
                            j:min(j+dx, ncshape[2])] == -1)] = -1

# save with stored Events/yr
pd.DataFrame(qPOT5.data).to_csv(
    "{}/Data/qPOT_{}epy_sans_mask.csv".format(mainfolder, events_per_year),
    index=False, header=None)

pd.DataFrame(qPOT5.mask).to_csv(
    "{}/Data/qPOT_{}epy_mask.csv".format(mainfolder, events_per_year),
    index=False, header=None)

# == dates of extreme events == 
extremes = []
qPOT6 = np.array(qPOT5.data)

for tt in range(0, ncshape[0]):
    if tt % 100 == 0:
        print("tt = " + str(tt))
    flow_slice = ncin.variables['dmflow'][tt,:,:]
    inun = np.sum((flow_slice > qPOT6) & (qPOT6 > -1))
    if inun > int(inun_thres*np.sum(qPOT6 > -1)):
        print(str(tt) + ": " + str(inun) + " / " + str(np.sum(qPOT6 > -1)))
        extremes.append([tt, inun])
        
extremes = pd.DataFrame(extremes, columns=['Day', 'Inun'])
        
pd.DataFrame(extremes).to_csv(
    "{}/Data/extremedates_{}inun_{}epy.csv".format(
        mainfolder, int(inun_thres*100), events_per_year),
    index=False, header=True)

ncin.close()